src/System.IO.Pipelines/MemoryPool.cs

```diff
@@ -0,0 +1,243 @@
+        /// </summary>
+        private const int _blockCount = 32;
+
+        /// <summary>
+        /// 4096 - 64 gives you a blockLength of 4032 usable bytes per block.
+        /// </summary>
+        private const int _blockLength = _blockStride - _blockUnused;
+
+        /// <summary>
+        /// Max allocation block size for pooled blocks, 
+        /// larger values can be leased but they will be disposed after use rather than returned to the pool.
+        /// </summary>
+        public const int MaxPooledBlockLength = _blockLength;
+
+        /// <summary>
+        /// 4096 * 32 gives you a slabLength of 128k contiguous bytes allocated per slab
+        /// </summary>
+        private const int _slabLength = _blockStride * _blockCount;
+
+        /// <summary>
+        /// Thread-safe collection of blocks which are currently in the pool. A slab will pre-allocate all of the block tracking objects
+        /// and add them to this collection. When memory is requested it is taken from here first, and when it is returned it is re-added.
+        /// </summary>
+        private readonly ConcurrentQueue<MemoryPoolBlock> _blocks = new ConcurrentQueue<MemoryPoolBlock>();
+
+        /// <summary>
+        /// Thread-safe collection of slabs which have been allocated by this pool. As long as a slab is in this collection and slab.IsActive, 
+        /// the blocks will be added to _blocks when returned.
+        /// </summary>
+        private readonly ConcurrentStack<MemoryPoolSlab> _slabs = new ConcurrentStack<MemoryPoolSlab>();
+
+        /// <summary>
+        /// This is part of implementing the IDisposable pattern.
+        /// </summary>
+        private bool _disposedValue = false; // To detect redundant calls
+
+        private Action<MemoryPoolSlab> _slabAllocationCallback;
+
+        private Action<MemoryPoolSlab> _slabDeallocationCallback;
+
+        public OwnedMemory<byte> Lease(int size)
+        {
+            if (size > _blockLength)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException_BufferRequestTooLarge(_blockLength);
+            }
+
+            var block = Lease();
+            block.Initialize();
+            return block;
+        }
+
+        public void RegisterSlabAllocationCallback(Action<MemoryPoolSlab> callback)
+        {
+            _slabAllocationCallback = callback;
+        }
+
+        public void RegisterSlabDeallocationCallback(Action<MemoryPoolSlab> callback)
+        {
+            _slabDeallocationCallback = callback;
+        }
+
+        /// <summary>
+        /// Called to take a block from the pool.
+        /// </summary>
+        /// <returns>The block that is reserved for the called. It must be passed to Return when it is no longer being used.</returns>
+#if DEBUG
+        private MemoryPoolBlock Lease(
+            [CallerMemberName] string memberName = "",
+            [CallerFilePath] string sourceFilePath = "",
+            [CallerLineNumber] int sourceLineNumber = 0)
```

Comment:
We'll need to figure out a way to get these diagnostics without being in debug mode. But these are fine for now